polkadot/node/network/collator-protocol/src/validator_side_experimental/state.rs

```diff
@@ -14,19 +14,510 @@
+	) {
+		let outcome = self
+			.peer_manager
+			.try_accept_connection(
+				sender,
+				peer_id,
+				PeerInfo { version, state: PeerState::Connected },
+			)
+			.await;
+
+		match outcome {
+			TryAcceptOutcome::Added => {},
+			TryAcceptOutcome::Replaced(other_peers) => {
+				gum::trace!(
+					target: LOG_TARGET,
+					"Peer {:?} replaced the connection slots of other peers: {:?}",
+					peer_id,
+					&other_peers
+				);
+				self.collation_manager.remove_peers(other_peers);
+			},
+			TryAcceptOutcome::Rejected => {
+				gum::debug!(
+					target: LOG_TARGET,
+					?peer_id,
+					"Peer connection was rejected. Going to disconnect",
+				);
+			},
+		}
+	}
+
+	/// Handle a peer disconnection.
+	pub async fn handle_peer_disconnected(&mut self, peer_id: PeerId) {
+		gum::trace!(
+			target: LOG_TARGET,
+			?peer_id,
+			"Peer disconnected",
+		);
+
+		self.peer_manager.disconnected(&peer_id);
+
+		self.collation_manager.remove_peers([peer_id].into_iter().collect());
+	}
+
+	/// Handle a peer's declaration message.
+	pub async fn handle_declare<Sender: CollatorProtocolSenderTrait>(
+		&mut self,
+		sender: &mut Sender,
+		peer_id: PeerId,
+		para_id: ParaId,
+	) {
+		if !self.peer_manager.declared(sender, peer_id, para_id).await {
+			self.collation_manager.remove_peers([peer_id].into_iter().collect());
+		}
+	}
+
+	/// Handle our view update.
+	pub async fn handle_our_view_change<Sender: CollatorProtocolSenderTrait>(
+		&mut self,
+		sender: &mut Sender,
+		new_view: OurView,
+	) -> FatalResult<()> {
+		gum::trace!(
+			target: LOG_TARGET,
+			?new_view,
+			"Handling our view change",
+		);
+		let old_assignments = self.collation_manager.assignments();
+
+		self.collation_manager.view_update(sender, &self.keystore, new_view).await;
+
+		let new_assignments = self.collation_manager.assignments();
+		gum::trace!(
+			target: LOG_TARGET,
+			?old_assignments,
+			?new_assignments,
+			"Old assignments vs new assignments",
+		);
+
+		let maybe_disconnected_peers =
+			self.peer_manager.scheduled_paras_update(sender, new_assignments).await;
+
+		if !maybe_disconnected_peers.is_empty() {
+			gum::trace!(
+				target: LOG_TARGET,
+				?maybe_disconnected_peers,
+				"Disconnecting peers due to our view change",
+			);
+		}
+
+		self.collation_manager.remove_peers(maybe_disconnected_peers);
+
+		Ok(())
+	}
+
+	/// Handle a finalized block notification.
+	pub async fn handle_finalized_block<Sender: CollatorProtocolSenderTrait>(
+		&mut self,
+		sender: &mut Sender,
+		hash: Hash,
+		number: BlockNumber,
+	) -> FatalResult<()> {
+		gum::trace!(
+			target: LOG_TARGET,
+			?hash,
+			number,
+			"Processing new block finality notification",
+		);
+
+		// Process reputation bumps
+		if let Err(err) = self
+			.peer_manager
+			.update_reputations_on_new_finalized_block(sender, (hash, number))
+			.await
+		{
+			err.split()?.log();
+		}
+
+		// Process potential changes in the registered paras set. TODO: we need a new runtime API
```

Comment:
Is this really needed ?  

If some para is unregistered, then it will not make blocks anymore. We keep track of when para last made a block and just prune all paras that have been inactive for N days.

Reply:
we could, but that's more complicated IMO. Another thing to persist off-chain in order to approximate the on-chain info. I'd rather query the chain if it's easy enough

Reply:
I don't think there is something less complicated than querying a runtime API :D 

It's not just a field in a DB. If you stop the node and restart it sometime later, the DB will not have been updated for a while and would probably evict all of the reputations.
Also, for on-demand parachains this gets even more subtle, because they may be producing blocks very rarely

Reply:
Let's do this separately: https://github.com/paritytech/polkadot-sdk/issues/9053